### PR TITLE
gh-93896: allow enabling asyncio.Runner calls to asyncio.set_event_loop independantly to loop_factory

### LINKED ISSUE #93896: IsolatedAsyncioTestCase and asyncio.run no-longer call asyncio.set_event_loop
in https://github.com/python/cpython/pull/31799/files#diff-1f2ae0f6c6010caf9d5f1c80cd6033a796ffe2b60554f5df84f554f4a08e622b the calls to `asyncio.set_event_loop()` were removed which breaks aiohttp https://github.com/aio-libs/aiohttp/issues/6757 and `asyncio.SafeChildWatcher`

this looks like an intentional breaking change - and if it is should be documented

### PR DESCRIPTION
**Pull request by graingert**
Fixes #93896

* gh-93896

### CODE REVIEW

File: Lib/asyncio/runners.py
Comment (graingert) [inline Lib/asyncio/runners.py:24]: this opts asyncio.Runner out of the policy system and asyncio.run opts back in

File: Lib/asyncio/runners.py
Comment (graingert) [inline Lib/asyncio/runners.py]:
```suggestion
                events.set_event_loop(None)
```

File: Lib/asyncio/runners.py
Comment (graingert) [inline Lib/asyncio/runners.py]:
```suggestion
    a ProactorEventLoop will be started on Windows or a SelectorEventLoop will
```

File: Lib/unittest/async_case.py
Comment (graingert) [inline Lib/unittest/async_case.py]:
```suggestion
        runner = asyncio.Runner(
            debug=True,
            loop_factory=asyncio.new_event_loop,
            set_policy_loop=True,
        )
```

File: Lib/asyncio/runners.py
Comment (Bluenix2) [inline Lib/asyncio/runners.py:39]: If this is classified as a bug (clearly it is causing issues), then fixing it shouldn't be blocking. By only implementing this behind an argument we are introducing a new feature which means that this will only be added to 3.12 (and not 3.11).

I would vote for not adding this argument.

File: Lib/asyncio/runners.py
Comment (Bluenix2) [inline Lib/asyncio/runners.py:189]: This long sentence is quite wordy and without knowing the change I would probably not understand what it means.

I would probably suggest not making this change (referring to this very sentence). To most, setting the loop in the event loop policy is probably synonymous to actually creating and running a loop.

File: Misc/NEWS.d/next/Library/2022-06-21-13-08-46.gh-issue-93896.2BsYCX.rst
Comment (Bluenix2) [inline Misc/NEWS.d/next/Library/2022-06-21-13-08-46.gh-issue-93896.2BsYCX.rst]:
```suggestion
Restored missing ``set_event_loop()`` calls in ``asyncio.run()`` and ``AsyncTestCase`` accidentally removed in the 3.11 betas.
```

**Review by Bluenix2 — CHANGES_REQUESTED** [review]

File: Lib/asyncio/runners.py
Comment (graingert) [reply Lib/asyncio/runners.py:39]: I think it's important to maintain the asyncio.Runner support for a high level policy-free use of asyncio, so that the policy system can be deprecated and removed

Comment (Bluenix2) [reply Lib/asyncio/runners.py:39]: By adding this option we make it *more difficult* to deprecate the policy system because then this option will need to be deprecated as well. Users who specify this option will have more to do to migrate.

Comment (graingert) [reply Lib/asyncio/runners.py:39]: How about I remove the flag from asyncio.Runner and have callers use a wrapper that mutates the policy loop?

```python
@contextmanager
def _with_policy_loop(loop):
    try:
        asyncio.set_event_loop(loop)
        yield
    finally:
        asyncio.set_event_loop(None)
```

```python
runner = asyncio.Runner()
cmgr = _with_policy_loop(runner.get_loop())
try:
    cmgr(runner.run)(corofn())
finally:
    cmgr(runner.close)()
```

> Users who specify this option will have more to do to migrate.

I'm proposing this `set_policy_loop` flag get deprecated at the same time as the policy system. I could add the flag as deprecated in this PR so users won't add it unless they already have more work to do to migrate from the policy system as a concept

Comment (Bluenix2) [reply Lib/asyncio/runners.py:39]: I don't think I understand the downside with enabling this by default. My arguments boil down to:

- With this flag disabled, we retain the current behaviour which causes issues. I would guess that most people don't fully understand what it does and will leave it to the default, then potentially run into these issues.

- As we've so far discussed, this flag is a fix for an issue which might soon no matter much because it is planned to be deprecated. With this option needing to be explicitly enabled, we introduce more things to deprecate. If, instead, this would be enabled by default then most would not need to change anything as it gets removed after the deprecation period.

I am not super confident in the event loop policy system, so I won't continue arguing if you don't agree with me after this comment - i'll defer to whenever a core dev takes a look at this issue.

Comment (graingert) [reply Lib/asyncio/runners.py:39]: > With this flag disabled, we retain the current behaviour which causes issues. I would guess that most people don't fully understand what it does and will leave it to the default, then potentially run into these issues.

It should be opt in for asyncio.Runner

Comment (graingert) [reply Lib/asyncio/runners.py:39]: yep: set_policy_loop defaults to False in `asyncio.Runner`: `def __init__(self, *, debug=None, loop_factory=None, set_policy_loop=False):`

people using the old `asyncio.run` get the old behaviour that uses the policy loop factory and sets the policy loop. People using the new `asyncio.Runner` get the new behaviour which does not use the policy loop factory or set the policy loop by default

File: Lib/asyncio/runners.py
Comment (graingert) [inline Lib/asyncio/runners.py]:
```suggestion
    def __init__(self, *, debug=None, loop_factory=None, set_policy_loop=None):
```

File: Lib/asyncio/runners.py
Comment (graingert) [inline Lib/asyncio/runners.py]:
```suggestion
        if set_policy_loop is None and loop_factory is None:
            # default to setting the policy loop, if the loop_factory is not set
            self._set_policy_loop = True
        else:
            self._set_policy_loop = set_policy_loop
```

**Comment by graingert** [comment]
I've merged the changes from https://github.com/python/cpython/pull/94593 so I'm now proposing this as a feature change instead of a bug fix

File: Lib/asyncio/runners.py
Comment (graingert) [inline Lib/asyncio/runners.py:145]: @kumaraditya303 it looks like your PR is missing some calls to set_event_loop here

Comment (kumaraditya303) [reply Lib/asyncio/runners.py:145]: `asyncio.Runner` should be used a context manager and its close method removes the set event loop so not required here.

Comment (graingert) [reply Lib/asyncio/runners.py:145]: the issue being when Runner.run finishes the loop should be unset

```python
with Runner() as runner1, Runner() as runner2:
    runner1.run(fn())  # calls set_event_loop
    runner2.run(fn())  # also calls set_event_loop
    # I'd expect the event loop to be unset here
```

Comment (kumaraditya303) [reply Lib/asyncio/runners.py:145]: >  I'd expect the event loop to be unset here

That happens only if you supply your `loop_factory` otherwise for existing code, its better to keep it set for existing code not using `asyncio.Runner`.

Comment (kumaraditya303) [reply Lib/asyncio/runners.py:145]: I would expect the event loop to be set until the context manager exit which is what we have currently.

Comment (graingert) [reply Lib/asyncio/runners.py:145]: Ah I see what you mean

**Comment by graingert** [comment]
I'll close this PR for now. It might make sense to re-introduce the `set_policy_loop` kwarg for 3.12 if the policy system is deprecated

**Pull request closed without merging**
